app/forms/otp_verification_form.rb

```diff
@@ -1,31 +1,65 @@
 class OtpVerificationForm
+  include ActiveModel::Model
+
+  validates :code, presence: true
+  validate :validate_code_length
+  validate :validate_code_matches_format
+  validate :validate_user_otp_presence
+  validate :validate_user_otp_expiration
+  validate :validate_code_equals_user_otp
+
   def initialize(user, code)
     @user = user
     @code = code
   end
 
   def submit
+    success = valid?
+
+    user.clear_direct_otp
+
     FormResponse.new(
-      success: valid_direct_otp_code?,
+      success: success,
+      errors: errors,
       extra: extra_analytics_attributes,
+      serialize_error_details_only: true,
     )
   end
 
   private
 
   attr_reader :code, :user
 
-  def valid_direct_otp_code?
-    return false unless code.match? pattern_matching_otp_code_format
-    user.authenticate_direct_otp(code)
+  def validate_code_length
+    return if code&.size == TwoFactorAuthenticatable::DIRECT_OTP_LENGTH
+    errors.add(:code, :code_incorrect_length, type: :code_incorrect_length)
+  end
+
+  def validate_code_matches_format
+    return if code&.match?(/^[0-9]+/i)
+    errors.add(:code, :code_pattern_mismatch, type: :code_pattern_mismatch)
+  end
+
+  def validate_user_otp_presence
+    return if user.direct_otp.present?
+    errors.add(:code, :user_otp_missing, type: :user_otp_missing)
+  end
+
+  def validate_user_otp_expiration
+    return if !otp_expired?
+    errors.add(:code, :user_otp_expired, type: :user_otp_expired)
   end
 
-  def pattern_matching_otp_code_format
-    /\A[0-9]{#{otp_code_length}}\z/i
```

Comment:
Previously, we enforced that the given code was numeric here, but then proceeded to do the comparison using Crockford Base32 normalization. I think a reasonable expectation is that we should not be receiving non-numeric codes for validation due to client-side form validation, so I don't know that we need to do the normalization tolerances at all. If / when this form would be used for validating alphanumeric codes (e.g. IdV phone verification), we could revisit?

I removed Crockford Base32 normalization as part of 383b435c0a773a4aa7243ff860294f0426188076